File: javascript/src/algorithms/cameltokebab.js
import { benchmark, print_recap } from "../tools.js";

// Premise :
// You receive a string list of word
// You kebab them, assuming they are camelCase



class camel_to_kebab_t {
    constructor() {
        // this.iterations = 1000000;
        this.iterations = 1;
    }

    method_0(words) {
        const regex = /[A-Z]+(?![a-z])|[A-Z]/g;

        return words.map(word => word.replace(regex, ($, pos) => {
            return `${(pos > 0) ? "-" : ""}${$.toLowerCase()}`;
        }));
    }

    method_1(words) {
        const regex = /([a-z])([A-Z])/g;
        return words.map(word => word.replace(regex, "$1-$2").toLowerCase());
    }

    method_2(words) { // Fastest (Template literals)
        const reg_find_maj = /[A-Z]+(?![a-z])|[A-Z]/g;
        const to_kebab = (strings, ...values) => {
            let str = values.reduce((final_string, val, idx) => {
                return `${final_string}${val}${strings[idx]}`;
            });

            return str.replace(reg_find_maj, (match, offset) => {
                return `${(offset > 0) ? "-" : ""}${match.toLowerCase()}`;
            });
        };

        return words.map(word => to_kebab`${word}`);
    }

    run_batch() {
        const input = ["helloWorld", "clippyTheAssistant", "BeanBurrito", "tastySalamanderA"];

        let res = {};
        res[0] = benchmark(this.method_0, input, this.iterations);
        res[1] = benchmark(this.method_1, input, this.iterations);
        res[2] = benchmark(this.method_2, input, this.iterations);

        print_recap(res);
    }
}

let camel_to_kebab = new camel_to_kebab_t();

export { camel_to_kebab };

/*
What happens in method 0?
const regex = /[A-Z]+(?![a-z])|[A-Z]/g;

In this regex we find : [A-Z]+(?![a-z])
This means, return every occurence of [A-Z] which is followed by an occurence of [a-z]
The negative lookahead (?!expr) simply means that we check were it is possible, but don't
actually return the second lowercase charater, only the first upper case character.

/g option simply means that we don't want to stop the operation after the first occurence,
but that we desire to keep going afterward and repeat the search

Then we have : |([A-Z])
Meaning OR any upper case letter, and this can only happen if a potential match
couldn't be done by the first test on the left of the OR operator

Finally, depending if we it's the first character or not, we just want to lowercase
the match, or add a - kebab stick.

We do so using the function as an argument in the replace string function, it will go
through this with every match

First, we filter based on the second argument of the functor, "offset"
this is the location in the string where the occurence was found, so in our case, 
if offset === 0, this means it's the first character, and that we do not want to add
a kebab stick, so we simply lowercase the match and return that.
If offset > 0, this means the occurence needs a kebabstick right before the lowercase
version of the match.


What happens in method 1?
const regex = /([a-z])([A-Z])/g;

This regex is much simpler, we simply return a match for every lowercase character
immediately followed by an uppercase character, this way if the first char is a maj,
it won't be given a kebab stick.

In the replace function, we use an expression to describe what to do with every match.
"$1-$2"
Here, $1 means everything on the left of the match, and $2 means everything on the right.
So we simply add a kebab stick in between this way

Finally, we lowercase the whole string.



*/

File: python/src/algorithms/camel_to_kebab.py
from ..benchmark import execute_test, print_report
import re

class camel_to_kebab_t(object):
    def __init__(self):
        self.iterations = 500000
        # self.iterations = 1

    def method_0(self, words):
        res = []

        rgx = re.compile(r'[A-Z]+(?![a-z])|[A-Z]')
        def convert_func(match):
            replacement = match.group()
            if (match.start() != 0):
                replacement = f"-{replacement}"
            return replacement.lower()

        for word in words:
            kebab = re.sub(rgx, convert_func, word)
            res.append(kebab)

        return res


    def method_0g(self, words):
        rgx = re.compile(r'[A-Z]+(?![a-z])|[A-Z]')
        def convert_func(match):
            replacement = match.group()
            if (match.start() != 0):
                replacement = f"-{replacement}"
            return replacement.lower()

        return [re.sub(rgx, convert_func, word) for word in words]

    # Prefered
    def method_0gl(self, words):
        rgx = re.compile(r'[A-Z]+(?![a-z])|[A-Z]')

        convert_func = lambda match: f"{'-' if match.start() != 0 else ''}{match.group().lower()}"

        return [re.sub(rgx, convert_func, word) for word in words]


    def method_1(self, words):
        res = []

        rgx = re.compile(r'(?<!^)(?=[A-Z])')

        for word in words:
            kebab = re.sub(rgx, "-", word).lower()
            res.append(kebab)

        return res

    # Fastest
    def method_1g(self, words):
        rgx = re.compile(r'(?<!^)(?=[A-Z])')
        return [re.sub(rgx, "-", word).lower() for word in words]

    def method_2(self, words):
        rgx = re.compile(r'(?<=[a-z])(?=[A-Z]|\d)|(?<=[A-Z]|\d)(?=[A-Z]{1}(?!$))(?!([A-Z]|\d){2,})|(?<=[A-Z])(?=\d{2,})|(?<=\d)(?=[A-Z]|[a-z])|(?<=[A-Z]|[a-z])(?=\d)')
        return[re.sub(rgx, "-", word).lower() for word in words]

    def run_batch(self):
        words = """helloWorld
                clippyTheAssistant
                BeanBurrito
                tastySalamanderA
                Los5000GatitosMuchoGrazie
                _1000SpicyHUDJuice80
                HELLO5000
                Hello5000
                HELLO
                H300
                5000A
                1545A158BOBTheBUILDER8
                0a1a5000heyHello
                _8000
                8000O
                8000A7000"""
        words = words.splitlines()
        words = [word.strip() for word in words]

        res = []
        res.append(execute_test("0", self.method_0, words, self.iterations))
        res.append(execute_test("0g", self.method_0g, words, self.iterations))
        res.append(execute_test("0gl", self.method_0gl, words, self.iterations))
        res.append(execute_test("1", self.method_1, words, self.iterations))
        res.append(execute_test("1g", self.method_1g, words, self.iterations))
        res.append(execute_test("2", self.method_2, words, self.iterations))

        print_report(res)


camel_to_kebab = camel_to_kebab_t()
